hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreLock.java

```diff
@@ -0,0 +1,531 @@
+   *
+   * @return The {@link LockInfo} for the found lock, or <code>null</code> if nothing found
+   */
+  private LockInfo findLock() throws LockException, InterruptedException {
+    Preconditions.checkArgument(
+        HiveVersion.min(HiveVersion.HIVE_2),
+        "Minimally Hive 2 HMS client is needed to find the Lock using the showLocks API call");
+    ShowLocksRequest showLocksRequest = new ShowLocksRequest();
+    showLocksRequest.setDbname(databaseName);
+    showLocksRequest.setTablename(tableName);
+    ShowLocksResponse response;
+    try {
+      response = metaClients.run(client -> client.showLocks(showLocksRequest));
+    } catch (TException e) {
+      throw new LockException(e, "Failed to find lock for table %s.%s", databaseName, tableName);
+    }
+    for (ShowLocksResponseElement lock : response.getLocks()) {
+      if (lock.getAgentInfo().equals(agentInfo)) {
+        // We found our lock
+        return new LockInfo(lock.getLockid(), lock.getState());
+      }
+    }
+
+    // Not found anything
+    return null;
+  }
+
+  private void unlock(Optional<Long> lockId) {
+    Long id = null;
+    try {
+      if (!lockId.isPresent()) {
+        // Try to find the lock based on agentInfo. Only works with Hive 2 or later.
+        if (HiveVersion.min(HiveVersion.HIVE_2)) {
+          LockInfo lockInfo = findLock();
+          if (lockInfo == null) {
+            // No lock found
+            LOG.info("No lock found with {} agentInfo", agentInfo);
+            return;
+          }
+
+          id = lockInfo.lockId;
+        } else {
+          LOG.warn("Could not find lock with HMSClient {}", HiveVersion.current());
+          return;
+        }
+      } else {
+        id = lockId.get();
+      }
+
+      doUnlock(id);
+    } catch (InterruptedException ie) {
+      if (id != null) {
+        // Interrupted unlock. We try to unlock one more time if we have a lockId
+        try {
+          Thread.interrupted(); // Clear the interrupt status flag for now, so we can retry unlock
+          LOG.warn("Interrupted unlock we try one more time {}.{}", databaseName, tableName, ie);
+          doUnlock(id);
+        } catch (Exception e) {
+          LOG.warn("Failed to unlock even on 2nd attempt {}.{}", databaseName, tableName, e);
+        } finally {
+          Thread.currentThread().interrupt(); // Set back the interrupt status
+        }
+      } else {
+        Thread.currentThread().interrupt(); // Set back the interrupt status
+        LOG.warn("Interrupted finding locks to unlock {}.{}", databaseName, tableName, ie);
+      }
+    } catch (Exception e) {
+      LOG.warn("Failed to unlock {}.{}", databaseName, tableName, e);
+    }
+  }
+
+  private void doUnlock(long lockId) throws TException, InterruptedException {
+    metaClients.run(
+        client -> {
+          client.unlock(lockId);
+          return null;
+        });
+  }
+
+  private static synchronized void initTableLevelLockCache(long evictionTimeout) {
```

Comment:
Nit: can we put this on top or bottom of method?  (as its static)

Reply:
Also is it worth wrapping this around double-check locking to avoid synchronizing here, for performance?  Although I guess acquire JVM lock already synchronizes the thread.  Up to you

Reply:
Moved.

I kept it `synchronized` - I think concurrent commits in the same JVM will not be a blocker for a while.

Reply:
Thanks for move.  Actually it may be worth to do a double-check locking?  (I mean, make the commitLockCache a volatile, keep synchronzied but move under commitLockCache != null check)  I see it done a lot in other parts of Iceberg code.

The optimization being, except first time, each thread wont unnecessarily need to enter a barrier every time it passes the method?  The way it works now, it looks like each thread should not need to enter a barrier, unless they are multiple threads interested in the same table?

Reply:
Done

Reply:
Thanks for the idea!
This is fixed now.